ddtrace/context.py

```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
         # type: () -> str
         if self.trace_id is None or self.span_id is None:
             return ""
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        if tp:
+            # grab the original traceparent trace id, not the converted value
+            trace_id = tp.split("-")[1]
+        else:
+            trace_id = "{:032x}".format(self.trace_id)
 
         sampled = 1 if self.sampling_priority and self.sampling_priority > 0 else 0
-        return "00-{:032x}-{:016x}-{:02x}".format(self.trace_id, self.span_id, sampled)
+        return "00-{}-{:016x}-{:02x}".format(trace_id, self.span_id, sampled)
+
+    @property
+    def _tracestate(self):
+        # type: () -> str
+        dd = ""
+        if self.sampling_priority:
+            dd += "s:{};".format(self.sampling_priority)
+        if self.dd_origin:
+            dd += "o:{};".format(self.dd_origin)
+        sampling_decision = self._meta.get(SAMPLING_DECISION_TRACE_TAG_KEY)
+        if sampling_decision:
+            # replace characters ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            dd += "t.dm:{};".format(re.sub(r",|=|[^\x20-\x7E]+", "_", sampling_decision))
+        # since this can change, we need to grab the value off the current span
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd += "t.usr.id:{};".format(self._meta.get(USER_ID_KEY))
+
+        # grab all other _dd.p values out of meta since we need to propagate all of them
+        for k, v in self._meta.items():
+            if isinstance(k, str) and k.startswith("_dd.p") and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]:
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",|=|[^\x20-\x7E]+", "_", k)), re.sub(r",|;|:|[^\x20-\x7E]+", "_", v)
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
+        # remove final ;
```

Comment:
Can we avoid adding the final `;` in the first place

Reply:
I'm not sure how we could, since we don't know when the final addition of a string is going to be added. I guess one way we could do it is by having each string start with a `;`. Let me try that and see if it works.

Reply:
ah I don't think it would work, as we can't count on the string added before each string to have the `;`.  For example here we can't count on there being a `sampling_priority` value and if there was only an `origin` value
 we'd have `dd=;o:rum` for example, which wouldn't be acceptable.
 ```
        if self.sampling_priority:
            dd += "s:{};".format(self.sampling_priority)
        if self.dd_origin:
            dd += "o:{};".format(self.dd_origin)
```

Reply:
Strings are immutable in python so removing a character will generate a new string. It would be a slight performance improvement if we could avoid this but that's not a big deal. We could enumerate over the items in the dictionary and for the last item we can avoid adding ';'.  This logic change would make the for loop harder to read.

tldr: ignore my original suggestion